### PR TITLE
feat: migrate to TypeScript

### PR DESCRIPTION
**Pull request by kentcdodds**
**What**: migrate to TypeScript

**Why**: TypeScript is great

**How**: Upgrade kcd-scripts and .ts all the things

**Checklist**:

- [ ] Documentation N/A
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/__tests__/index.ts [hunk 1]
```diff
@@ -25,7 +25,8 @@ test("useDeepCompareEffectNoCheck don't throw an error if using it with an array
     ),
   ).not.toThrow()
   expect(console.error).toHaveBeenCalledTimes(0)
-  console.error.mockRestore()
+  // @ts-expect-error no idea
+  console.error.mockRestore() // eslint-disable-line
```
Comment (kentcdodds) [inline src/__tests__/index.ts:29]: I'm not sure how to deal with this.

Comment (rbusquet) [reply src/__tests__/index.ts:29]:
```ts
const consoleMock = jest.spyOn(console, 'error', ...)
// ...
consoleMock.mockRestore();
```

File: src/index.ts [hunk 2]
```diff
@@ -1,7 +1,12 @@
-import React from 'react'
+import * as React from 'react'
 import {dequal as deepEqual} from 'dequal'
 
-function checkDeps(deps) {
+type UseEffectParams = Parameters<typeof React.useEffect>
+// yes, I know it's void, but I like what this communicates about
+// the intent of these functions: It's just like useEffect
+type UseEffectReturn = ReturnType<typeof React.useEffect>
```
Comment (rbusquet) [inline src/index.ts:7]: this is nice :D

File: src/__tests__/index.ts [hunk 3]
```diff
@@ -96,7 +97,7 @@ test('useDeepCompareEffect does NOT work with manipulation', () => {
 
 test('useDeepCompareEffect works with deep object similarities/differences', () => {
   const callback = jest.fn()
-  let deps = [{a: {b: {c: 'd'}}}]
+  let deps: Array<Record<string, unknown>> = [{a: {b: {c: 'd'}}}]
```
Comment (Aprillion) [inline src/__tests__/index.ts:100]: what TS complained about without an explicit type here?

Comment (kentcdodds) [reply src/__tests__/index.ts:100]: Yeah, it complained because I added a property to the object.

**Comment by kentcdodds** [comment]
@all-contributors please add @Aprillion for review

**Comment by kentcdodds** [comment]
@all-contributors please add @rbusquet for review

File: src/__tests__/index.ts (hunk 3 above)
Comment (Aprillion) [reply src/__tests__/index.ts:99]: oh, right... hm, the inference choices about what is mutable and what not can still surprise me - I guess an array of objects makes sense that the objects should have the same shape by default

File: src/index.ts [hunk 4]
```diff
@@ -28,15 +35,21 @@ function useDeepCompareMemoize(value) {
...
-function useDeepCompareEffect(callback, dependencies) {
+function useDeepCompareEffect(
+  callback: EffectCallback,
+  dependencies: DependencyList,
+): UseEffectReturn {
   if (process.env.NODE_ENV !== 'production') {
     checkDeps(dependencies)
   }
   // eslint-disable-next-line react-hooks/exhaustive-deps
   React.useEffect(callback, useDeepCompareMemoize(dependencies))
```
Comment (MichaelDeBoey) [inline src/index.ts:46]: @kentcdodds Why not return `React.useEffect` in this case? 🤔

Your return type states that it's the same?

File: src/index.ts [hunk 5]
```diff
@@ -28,15 +35,21 @@ function useDeepCompareMemoize(value) {
...
 }
 
-export function useDeepCompareEffectNoCheck(callback, dependencies) {
+export function useDeepCompareEffectNoCheck(
+  callback: EffectCallback,
+  dependencies: DependencyList,
+): UseEffectReturn {
   // eslint-disable-next-line react-hooks/exhaustive-deps
   React.useEffect(callback, useDeepCompareMemoize(dependencies))
```
Comment (MichaelDeBoey) [inline src/index.ts:54]: Same

File: src/index.ts (hunk 4 above)
Comment (Aprillion) [reply src/index.ts:46]: well, if it returns undefined, not returning is the same think as returning React.useEffect, no?

Comment (MichaelDeBoey) [reply src/index.ts:46]: It's actually the same indeed, but I think it's more expressing what's happening to return `React.useEffect`

Comment (kentcdodds) [reply src/index.ts:46]: Good point. Done 👍

**Pull request merged**